grpc-sys/src/lib.rs

```diff
@@ -379,24 +379,144 @@ pub union GrpcSliceData {
 pub enum GrpcSliceRefCount {}
 
 #[repr(C)]
-#[derive(Clone, Copy)]
 pub struct GrpcSlice {
     ref_count: *mut GrpcSliceRefCount,
     data: GrpcSliceData,
 }
 
+impl GrpcSlice {
+    pub fn with_capacity(len: usize) -> Self {
+        unsafe { grpc_slice_malloc(len) }
+    }
+
+    pub fn len(&self) -> usize {
+        unsafe { grpcwrap_slice_length(self) }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn range_from(&self, offset: usize) -> &[u8] {
+        unsafe {
+            let mut len = 0;
+            let ptr = grpcwrap_slice_raw_offset(self, offset, &mut len);
+            slice::from_raw_parts(ptr as _, len)
+        }
+    }
+}
+
+/// Increase the ref count of the slice when cloning
+impl Clone for GrpcSlice {
+    fn clone(&self) -> Self {
+        unsafe { grpcwrap_slice_ref(self) }
+    }
+}
+
+impl Default for GrpcSlice {
+    fn default() -> Self {
+        unsafe { grpc_empty_slice() }
+    }
+}
+
+/// Decrease the ref count of the slice when dropping
+impl Drop for GrpcSlice {
+    fn drop(&mut self) {
+        unsafe {
+            grpcwrap_slice_unref(self);
+        }
+    }
+}
+
+impl<'a> From<&'a [u8]> for GrpcSlice {
+    fn from(data: &'a [u8]) -> Self {
+        unsafe { grpc_slice_from_copied_buffer(data.as_ptr() as _, data.len()) }
+    }
+}
+
 #[repr(C)]
 pub union GrpcByteBufferReaderCurrent {
     index: c_uint,
 }
 
 #[repr(C)]
 pub struct GrpcByteBufferReader {
-    pub buffer_in: *mut GrpcByteBuffer,
-    pub buffer_out: *mut GrpcByteBuffer,
+    pub buffer_in: *mut GrpcByteBufferRaw,
+    pub buffer_out: *mut GrpcByteBufferRaw,
     current: GrpcByteBufferReaderCurrent,
 }
 
+impl GrpcByteBufferReader {
+    pub fn len(&self) -> usize {
+        unsafe { grpc_byte_buffer_length(self.buffer_out) }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn next_slice(&mut self) -> GrpcSlice {
+        unsafe {
+            let mut slice = Default::default();
+            let code = grpc_byte_buffer_reader_next(self, &mut slice);
+            debug_assert_ne!(code, 0);
+            slice
+        }
+    }
+}
+
+impl<'a> From<&'a mut GrpcByteBuffer> for GrpcByteBufferReader {
+    fn from(src: &'a mut GrpcByteBuffer) -> Self {
```

Comment:
Why not let `GrpcByteBufferReader` holds `GrpcByteBuffer` instead of some raw pointers? So we don't need `_buf` in the `MessageReader`. And so I think we can merge `GrpcByteBuffer` and `Raw`.

Reply:
No, we need `Raw` version. In some existing code, we need to hold it as a reference and we cannot destruct it.
Also, `GrpcByteBufferReader` is a struct from C side, it's probably a bad idea to change its structure.

Reply:
Also, we always operate `GrpcByteBuffer` as a pointer. If we want to RAII it, we need to create the complete struct in Rust, which will bring extra work (this struct depend on other structs, and recurse).

Reply:
So why we need `GrpcByteBuffer`? And if GrpcByteBuffer has only 1 pointer field, their memory layout are exactly equal.